File: app/src/main/java/ru/roscha_akademii/medialib/common/AndroidModule.java
package ru.roscha_akademii.medialib.common;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;

import com.squareup.leakcanary.RefWatcher;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AndroidModule {
    private final MediaLibApplication application;
    private final RefWatcher refWatcher;

    public AndroidModule(MediaLibApplication application, RefWatcher refWatcher) {
        this.application = application;
        this.refWatcher = refWatcher;
    }

    @Provides
    @Singleton
    Context providesContext() {
        return application;
    }

    @Provides
    @Singleton
    RefWatcher providesRefWatcher() {
        return refWatcher;
    }
    
    @Provides
    @Singleton
    Resources providesResources() {
        return application.getResources();
    }

    @Provides
    @Singleton
    SharedPreferences providesSharedPreferences(Context context) {
        return context.getSharedPreferences("app", Context.MODE_PRIVATE);
    }

}

File: app/src/main/java/ru/roscha_akademii/medialib/common/AndroidModule.kt
package ru.roscha_akademii.medialib.common

import android.app.DownloadManager
import android.content.ContentResolver
import android.content.Context
import android.content.SharedPreferences
import android.content.res.Resources

import com.squareup.leakcanary.RefWatcher
import com.squareup.picasso.Picasso

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import org.greenrobot.eventbus.EventBus

@Module
open class AndroidModule(private val application: MediaLibApplication,
                         private val refWatcher: RefWatcher) {

    @Provides
    @Singleton
    internal fun providesContext(): Context = application

    @Provides
    @Singleton
    internal fun providesRefWatcher() = refWatcher

    @Provides
    @Singleton
    internal fun providesResources() = application.resources

    @Provides
    @Singleton
    internal open fun providesDownloadManager()
            = application.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager

    @Provides
    @Singleton
    internal fun providesContentResolver() = application.contentResolver

    @Provides
    @Singleton
    internal fun providesSharedPreferences(context: Context)
            = context.getSharedPreferences("app", Context.MODE_PRIVATE)

    @Provides
    @Singleton
    internal fun providesPicasso(context: Context) = Picasso.with(context)

    @Provides
    @Singleton
    internal fun providesEventBus() = EventBus.getDefault()
}
